HW0/P5/P5.txt

```diff
@@ -1 +1,57 @@
 # Your answers here.
+
+1. one cashier has too add 256 numbers, which means 255 additions.
+Therefore, 255 seconds.
+
+Answer : 255 sec
+
+
+2. total 8 coworkers will work together. So each person will have 256/8 = 32 bags.
+So adding up 32 bags will take 31 seconds.
+After that, each person has to add up what they've got with the coworker to get the total sum.
+If we think about tree structure, the height of the tree will be 3.
+Summing up each level will take 1 seconds,
+thereby resulting total time to be 31+3 = 34 seconds.
+
+Answer : 34 sec
+
+
+3. If there are infinite number of employees,
+then we could think of a tree that has 256 leaf nodes at the bottom.
+Since the height of the tree is log2(256) = 8,
+it will take 8 seconds.
+
+Answer : 8 sec
+
+
+4. in P5.png
+alone : N-1 sec
+infinite : log2(N) sec
+
+
+5. Again, when we draw a tree, the height would be 8 in total.
+So between each steps, communication has to be made.
+Therefore, 7 seconds will take additionally because of communication in total.
+So 8 sec (for calculation) + 7 sec (for communication) = 15 sec.
+
+Answer : 15 sec
+
+
+6. 1 bag at a time : In addition to the time to communicate, which is 16 sec,
+ 256 seconds will take to hand in 256 bags one by one.
+ Answer : 256 + 16 = 272 sec
+
+ 2 bags at a time : bags should be handed in 128 times (thus total 256 bags),
+ 128 sec will take additionally.
+ Answer : 128 + 15 = 143 sec
+
+ Better strategy : Give N bags to the first cashier, and then N-1 bags to second cashier,
+  and N-2 to third cashier and so on. Then at the moment when I give the last one bag to the last cashier,
+  the other cashiers would have done their summation.
+  To get the N, N + (N-1) + ... + 1 = N(N-1)/2 should be 256.
+  Such N is around 23. So until now it takes 23 seconds.
+  Then, N people should communicate and sum up.
+  Communication will take log2(32) (since 32 is the nearest number of power of 2 to 23) seconds for communication
+  and log2(32) seconds for summing up.
+  Therefore, 5 + 5 + 23 = 33 seconds in total.
+
```

Comment:
Clear explanation in general! I'm not too sure about #6, because I feel like communication and handing out the bag can be done concurrently.